src/chess/board.cc

```diff
@@ -982,8 +982,36 @@ void ChessBoard::SetFromFen(const std::string& fen, int* rule50_ply,
   string en_passant;
   int rule50_halfmoves;
   int total_moves;
-  fen_str >> board >> who_to_move >> castlings >> en_passant >>
-      rule50_halfmoves >> total_moves;
+  fen_str >> board;
+  if (fen_str.eof()) {
+    who_to_move = "w";
+  } else {
+    fen_str >> who_to_move;
+  }
+  if (fen_str.eof()) {
+    // Assume no castling rights. Other engines, e.g., Stockfish, assume kings
+    // and rooks on their initial rows can each castle with the outer-most rook.
+    // Our implementation currently supports 960 castling where white and black
+    // rooks have matching columns, so it's unclear which rights to assume.
+    castlings = "-";
+  } else {
+    fen_str >> castlings;
+  }
+  if (fen_str.eof()) {
+    en_passant = "-";
+  } else {
+    fen_str >> en_passant;
+  }
+  if (fen_str.eof()) {
+    rule50_halfmoves = 0;
+  } else {
+    fen_str >> rule50_halfmoves;
+  }
+  if (fen_str.eof()) {
+    total_moves = 1;
+  } else {
+    fen_str >> total_moves;
+  }
 
   if (!fen_str) throw Exception("Bad fen string: " + fen);
```

Comment:
Also I'm surprised that this works.. But as tests show that it does, it probably does. :)

I thought that when eof bit is set, the stream should return false.

Reply:
The unit test would fail when directly passing in a string with trailing space to `SetFromFen` but `ParseCommand` tokenizes and combines with a single space https://github.com/LeelaChessZero/lc0/blob/9b3c7992aef13a477b084fdaf3304fea5ed9e4b4/src/chess/uciloop.cc#L81-L87

Reply:
`operator bool` / `fail` are based on `failbit` and `badbit`: http://www.cplusplus.com/reference/ios/ios/fail/